### ISSUE TITLE
Allow to reset tab navigation stack on changing tabs.

### ISSUE THREAD

**Issue by ciekawy** [ionic-team/ionic-framework#17761]
# Feature Request

**Ionic version:**
[x] **4.x**

**Describe the Feature Request**
Allow to reset tab view stack on changing tabs.

Currently `ion-tabs` navigation on changing the tab to one selected in the past having some sub views opened brings back the the most recent view for this tab.

**Describe Preferred Solution**
in some cases restoring view stack may be confusing to the user - thus there should be an option to reset the tab to the root component.

also both `ionTabsWillChange` and `ionTabsDidChange` are called just one after another emitting same new tab while I'd expect to be able to know also the leaving tab (add `leavingTab` property to the event?). 

finally `ionTabsWillChange` is fired only after the new tab is already visible while I'd expect to get `ionTabsWillChange` before activating new tab and only `ionTabsWillChange` after  (this probably can be a separate bug)

**Describe Alternatives**

if there is a way to achieve this without side effect - would be good to add this information to the documentation.

**Related Code**

**Additional Context**

here is similar concern from another person on SO:
https://stackoverflow.com/questions/53687233/how-to-reset-route-stack-after-routing-away-from-tab-using-angular-ionic-4#comment96962743_53687233

**Comment by leoruhland** [comment]
I got a dirty solution while we don't have this feature on Ionic:

https://stackoverflow.com/a/55733165/5760235

**Comment by leoruhland** [comment]
Hey @ciekawy!

I think I have a temporary solution that does not seem to have any side-effects (yet).

On every tab in tabs.page.html:
```html
<ion-tab-button tab="tab1" (click)="openTab('tab1', $event)">
    <ion-label>Tab</ion-label>
</ion-tab-button>
```

On tabs.page.ts:
```ts
async openTab(tab: string, evt: MouseEvent) {
    const tabSelected = this.tabs.getSelected();
    evt.stopImmediatePropagation();
    evt.preventDefault();
    return tabSelected !== tab
      ? await this.navCtrl.navigateRoot(this.tabs.outlet.tabsPrefix + '/' + tab)
      : this.tabs.select(tab);
}
```

**Comment by jeffricks** [comment]
For anyone following, the workaround provided by @leoruhland worked great (thanks!) up through ionic 4.7.4 but seems to have stopped working with 8.0.0.  Has anyone upgraded to 8.0 and still have this working?

**Comment by ciekawy** [comment]
inspired by @leoruhland solution I've modified it a bit. Haven't checked though yet with Angular 8 - I believe with the use of `outlet.canGoBack` / `outlet.pop` there is a chance it will work also for 8:

```html
<ion-tabs #tabs>
    <ion-tab-button tab="tab1" (click)="handleTabClick($event)">
        <ion-label>Tab</ion-label>
    </ion-tab-button>
</ion-tabs> 
```

```typescript
// usual @Component() goes here
class TabsPage {
    @ViewChild('tabs') tabs: IonTabs;

    resetStackTabs = ['inbox', 'tasks'];
    handleTabClick = (event: MouseEvent) => {
        const { tab } = event.composedPath().find((element: any) =>
            element.tagName === 'ION-TAB-BUTTON') as EventTarget & { tab: string };
        // without checking resetStackTabs this will work for any tab
        if (this.resetStackTabs.includes(tab) &&
            this.tabs.outlet.canGoBack(1, tab)) {
            event.stopImmediatePropagation();

            // here we may need pop depth more than one if we handle deeper nav for a tab
            return this.tabs.outlet.pop(1, tab);
        }
    }
}
```

**Comment by viniciusmaran** [comment]
> inspired by @leoruhland solution I've modified it a bit. Haven't checked though yet with Angular 8 - I believe with the use of `outlet.canGoBack` / `outlet.pop` there is a chance it will work also for 8:
> 
> ```
> <ion-tabs #tabs>
>     <ion-tab-button tab="tab1" (click)="handleTabClick($event)">
>         <ion-label>Tab</ion-label>
>     </ion-tab-button>
> </ion-tabs> 
> ```
> 
> ```ts
> // usual @Component() goes here
> class TabsPage {
>     @ViewChild('tabs') tabs: IonTabs;
> 
>     resetStackTabs = ['inbox', 'tasks'];
>     handleTabClick = (event: MouseEvent) => {
>         const { tab } = event.composedPath().find((element: any) =>
>             element.tagName === 'ION-TAB-BUTTON') as EventTarget & { tab: string };
>         // without checking resetStackTabs this will work for any tab
>         if (this.resetStackTabs.includes(tab) &&
>             this.tabs.outlet.canGoBack(1, tab)) {
>             event.stopImmediatePropagation();
> 
>             // here we may need pop depth more than one if we handle deeper nav for a tab
>             return this.tabs.outlet.pop(1, tab);
>         }
>     }
> }
> ```

Worked for me. Thanks!

**Comment by fmendoza** [comment]
A modified version that handle any deep:

```js
export class TabsPage {

  @ViewChild(IonTabs) tabs: IonTabs;

  resetStackTabs = ['home', 'browse', 'cart', 'account'];

  handleTabClick = (event: MouseEvent) => {
    const { tab } = event.composedPath().find((element: any) =>
      element.tagName === 'ION-TAB-BUTTON') as EventTarget & { tab: string };

    let deep = 1;
    let canGoBack = false;

    const deepFn = () => {
      if (this.tabs.outlet.canGoBack(deep, tab)) {
        canGoBack = true;
        deep++;
        deepFn();
      }
    }

    deepFn();

    if (this.resetStackTabs.includes(tab) && canGoBack) {
      event.stopImmediatePropagation();
      return this.tabs.outlet.pop(deep - 1, tab);
    }
  }

  constructor() { }
}
```

**Comment by kennethhuang31** [comment]
> A modified version that handle any deep:
> 
> ```js
> export class TabsPage {
> 
>   @ViewChild(IonTabs) tabs: IonTabs;
> 
>   resetStackTabs = ['home', 'browse', 'cart', 'account'];
> 
>   handleTabClick = (event: MouseEvent) => {
>     const { tab } = event.composedPath().find((element: any) =>
>       element.tagName === 'ION-TAB-BUTTON') as EventTarget & { tab: string };
> 
>     let deep = 1;
>     let canGoBack = false;
> 
>     const deepFn = () => {
>       if (this.tabs.outlet.canGoBack(deep, tab)) {
>         canGoBack = true;
>         deep++;
>         deepFn();
>       }
>     }
> 
>     deepFn();
> 
>     if (this.resetStackTabs.includes(tab) && canGoBack) {
>       event.stopImmediatePropagation();
>       return this.tabs.outlet.pop(deep - 1, tab);
>     }
>   }
> 
>   constructor() { }
> }
> ```

Totally fixed my problem

**Comment by sstraus** [comment]
Hi, with Ionic 5 handleTabClick is not available anymore as event. Any idea on how to fix now?

**Comment by gaetancollaud** [comment]
Hi, with Ionic 5, my dirty solution is to directly use the router:
```typescript
  // Tabs that need reset when you click on them
  resetStackTabs = ['koalas'];

  @ViewChild('tabs') tabs: IonTabs;

  constructor(private router:Router) {
  }

  handleTabClick = (event: MouseEvent) => {
    const {tab} = event.composedPath().find((element: any) =>
      element.tagName === 'ION-TAB-BUTTON') as EventTarget & { tab: string };

    if (this.resetStackTabs.includes(tab)){
      this.router.navigate(['tabs/'+tab]);
    }
  };
```
```html
<ion-tab-button tab="koalas" (click)="handleTabClick( $event)">
```

**Comment by MarcioJrios** [comment]
@gaetancollaud that worked for me, thanks a lot!

**Comment by darron1217** [comment]
Is there any solution for React?

**Comment by liamdebeasi** [comment]
Hi everyone,

I have discussed this feature with the team. This feature goes against how we intend tabs to work, so we feel that it is not a good fit for Ionic.

In mobile applications (both Ionic and native), each tab is a standalone navigation stack. When you switch between tabs, the "leaving" navigation stack state is preserved. So if you are on `/tab1/child` and you tap the "Tab 2" button, later tapping the "Tab 1" button will bring you back to `/tab1/child`, not `/tab1`.

Additionally, items in one navigation stack should never interact with items in another navigation stack. This means that there should never be a link in Tab 1 that alters or changes the history of Tab 2.

These two practices form the basis for why we do not feel this is a good feature to add. The other constraint here is how much control Ionic has over the browser history. The browser does not let us programmatically rewrite parts of the browser history. We investigated doing this feature using `router.replace` as an alternative, but we found that it led to many unexpected behaviors when switching between tabs.

I am going to close this issue. Thank you!
